### ISSUE TITLE
Bip47 & Reusable Payment Codes version 3 & 4

### ISSUE THREAD

**Issue by yoshimo** [BlueWallet/BlueWallet#2883]
Since we are getting closer to TOR Support, it would be great to get additional support for privacy enhancing features.
Colloquial, PayNyms.
These are specified in Bip47  as implemented by Samourai and enhanced in https://github.com/OpenBitcoinPrivacyProject/rfc/blob/master/obpp-05.mediawiki
Having more wallets support this, would benefit the ecosystem in general.

**Comment by yoshimo** [comment]
150k Sats for an implementation

**Comment by yoshimo** [comment]
@Overtorment  could you add the bounty-label?

**Comment by librepgp** [comment]
Count me in for another 850k sats

**Comment by acantril** [comment]
Count me in for 1 million sats

**Comment by dav1dpgit** [comment]
In for 100k sats

**Comment by LaserHodl** [comment]
This could be huge. In for 100k sats.

**Comment by johnnywonder1** [comment]
I pledge 25,000 satoshis donated over Paynym upon completion.

**Comment by modl21** [comment]
count me in for a million sats

**Comment by RequestPrivacy** [comment]
Not a fan of iOS but it will help Bitcoin so 25,000 sats from me.

---
Total bounty: **4.250.000 sats**

**Comment by librepgp** [comment]
Another 250k from @Jorj_X_McKie on Twitter

https://twitter.com/Jorj_X_McKie/status/1494366711709159430?s=20&t=vKlwZaVyPHEg5kiKgPg4Xg

---

Total bounty: 4,500,000 SATs

**Comment by librepgp** [comment]
Member of Citadel Dispatch matrix group pledged another 1 million sats

<img width="561" alt="Screen Shot 2022-02-17 at 10 41 55" src="https://user-images.githubusercontent.com/45250003/154549073-4283270f-cad8-483f-9ece-67a209167cfc.png">

---

Total bounty: 5,500,000 SATs

**Comment by Ziya-Sadr** [comment]
count me in for 100k sats

**Comment by btcuserguide** [comment]
i'm in for 1,000,000

**Comment by BTclaudia** [comment]
I'm down for 500k SATs

**Comment by ion497** [comment]
Count me in for 200k sats

**Comment by librepgp** [comment]
Another 300k from [@Does_not_mattr](https://twitter.com/Does_not_mattr) on Twitter

https://twitter.com/Does_not_mattr/status/1495321573208858625

---

Total bounty: 8,100,000 SATs

**Comment by 0xAsteria** [comment]
200k from me

**Comment by jamesgmorgan** [comment]
I pledge 125K sats donated over paynym upon completion.

**Comment by librepgp** [comment]
Another 100k from [Meister Eder](https://twitter.com/Meister_Ancap) on Twitter

https://twitter.com/Meister_Ancap/status/1495473258258382850

---

Total bounty: 8,525,000 SATs

**Comment by diegoperezcham** [comment]
Count me in for 475K sats

**Comment by TheGuySwann** [comment]
Put me down for 1,000,000 sats to make it an even 10 👍🏻

**Comment by alaznem** [comment]
count me in for 40k sat

**Comment by FreddyPPFinklemeyerIII** [comment]
I’m in for 1,000,000 million sats

**Comment by librepgp** [comment]
Another 500k from [Isaiah Douglass, MBA, CFP](https://twitter.com/IDFinancial) on Twitter

https://twitter.com/IDFinancial/status/1495503166133215232

---

Total bounty: 11,500,040 SATs

**Comment by librepgp** [comment]
Another 300k from [AmBird](https://twitter.com/TurboSloth1869) on Twitter

https://twitter.com/TurboSloth1869/status/1495554918656233472

---

Total bounty: 11,800,040 SATs

**Comment by librepgp** [comment]
Another 100k from [Honk Honk MetaMick](https://twitter.com/metamick14) on Twitter

https://twitter.com/metamick14/status/1495850492978733062

---

Total bounty: 11,900,040 SATs

**Comment by pointbiz** [comment]
"bc1qq5y96828txk8lruha27l8gpkhtfc55u3f87s5q"
"JxylYiEmkCdFTxHd+u2MBfvZpH66oUhw8v+Dm5y7PMh4bGDOTYQZnxywHTYjGY/MjVJKLpzUVq0NAYxwtcojRCs="
"I, pointbiz aka BTCCuracao, pledge 19,500,000 sat bounty for Bluewallet github issue 2883. BIP47 v3 v4 reusable payment codes. The bounty to be paid after the first successful release of this feature as I deem it."

**Comment by karozagorus** [comment]
https://twitter.com/EricSirion/status/1494436823036272640

**Comment by karozagorus** [comment]
> > https://twitter.com/EricSirion/status/1494436823036272640
> 
> That's not how it works. Eric even says he realizes that later on the thread...

He never stated that apaprently.

**Comment by ncoelho** [comment]
Not sure why you are trying to fud bip47. Which is 7 year old bip with a lot of review.

Eric says he doesn't understand it. Which is natural if you didn't look deep enough.

https://twitter.com/ericsirion/status/1494633806069960707?s=21

**Comment by karozagorus** [comment]
> Not sure why you are trying to fud bip47. Which is 7 year old bip with a lot of review.
> 
> Eric says he doesn't understand it. Which is natural if you didn't look deep enough.
> 
> https://twitter.com/ericsirion/status/1494633806069960707?s=21

I'm just worried that Bluewallet's privacy might get compromised if this is allowed into it, especially since people are giving you monetary incentives now to add features.

**Comment by acantril** [comment]
That whole thread is embarrassing - imagine someone saying 'why bother with bitcoin or lightning, or i don't see the point in BTC/LN'

**Comment by karozagorus** [comment]
> That whole thread is embarrassing - imagine someone saying 'why bother with bitcoin or lightning, or i don't see the point in BTC/LN'

Agreed, its a cringe anti-LN stance they are doing.

**Comment by acantril** [comment]
> I'm just worried that Bluewallet's privacy might get compromised if this is allowed into it, especially since people are giving you monetary incentives now to add features.

Privacy compromised by a privacy feature?
and privacy of a product which has no privacy as it uses the BTC, by default 0 privacy.

**Comment by acantril** [comment]
Actually no, embarrassing isn't strong enough .. but i don't want to use the word i'm really thinking of.

**Comment by ncoelho** [comment]
> > Not sure why you are trying to fud bip47. Which is 7 year old bip with a lot of review.
> 
> > 
> 
> > Eric says he doesn't understand it. Which is natural if you didn't look deep enough.
> 
> > 
> 
> > https://twitter.com/ericsirion/status/1494633806069960707?s=21
> 
> 
> 
> I'm just worried that Bluewallet's privacy might get compromised if this is allowed into it, especially since people are giving you monetary incentives now to add features.

Bip47 is onchain, doesn't use 3rd parties or servers. So it is not possible to compromise bluewallet's privacy.

Users are putting a bounty, other users forked and are trying to get the bounty. There's zero monetary incentives to bluewallet.

**Comment by karozagorus** [comment]
> > > Not sure why you are trying to fud bip47. Which is 7 year old bip with a lot of review.
> > 
> > 
> > > 
> > 
> > 
> > > Eric says he doesn't understand it. Which is natural if you didn't look deep enough.
> > 
> > 
> > > 
> > 
> > 
> > > https://twitter.com/ericsirion/status/1494633806069960707?s=21
> > 
> > 
> > I'm just worried that Bluewallet's privacy might get compromised if this is allowed into it, especially since people are giving you monetary incentives now to add features.
> 
> Bip47 is onchain, doesn't use 3rd parties or servers. So it is not possible to compromise bluewallet's privacy.
> 
> Users are putting a bounty, other users forked and are trying to get the bounty. There's zero monetary incentives to bluewallet.

But doesn't the channel establishment require you to pay a fee for scamourai wallet?

**Comment by ncoelho** [comment]
> > > > 
> But doesn't the channel establishment require you to pay a fee for scamourai wallet?

No. As I said this is bitcoin onchain, there's samourai or servers involved. 

Nothing like learning something before commenting about it.

**Comment by ncoelho** [comment]
> > > >
> Well, since nobody verified this to be not leaking something I still don't trust it.

What do you mean? It literally is a bip (47) on bitcoin repository reviewed by some of the best cryptographers in the world.

You may not like samourai wallet implementation. But this has literally nothing to do with it.

**Comment by karozagorus** [comment]
> > > > >
> > 
> > 
> > Well, since nobody verified this to be not leaking something I still don't trust it.
> 
> What do you mean? It literally is a bip (47) on bitcoin repository reviewed by some of the best cryptographers in the world.
> 
> You may not like samourai wallet implementation. But this has literally nothing to do with it.

Well, it is very difficult to believe it because scamourai wallet's only goal is to co-opt bitcoin privacy and deanonimize users.

**Comment by ncoelho** [comment]
> > > > > >
> Well, it is very difficult to believe it because scamourai wallet's only goal is to co-opt bitcoin privacy and deanonimize users.

I don't understand why do you think this has anything to do with samourai wallet? This was not invented or purpose by them.

It is like saying you don't trust bip32, or bip84, which both samourai and also wasabi use... It is just a bip.

**Comment by RequestPrivacy** [comment]
This isn't Twitter. Stop spamming this issue with misleading suggestions. If you don't understand what a BIP is go and do some reading before making false claims.

**Comment by pointbiz** [comment]
In simple English, a Bitcoin address is something you give another person to request a payment from them. The act of giving the address to someone identifies the address belongs to you. That same address is used in the blockchain and anyone can see for free that YOU are receiving transactions.

Reusable payment codes is a new type of address (P address). When you give someone your P address you are identified with it. But this P address never is put in the blockchain. The P address is used to make a.new address that only the sender and receiver know who it belongs to. That new address is used in the blockchain.

This is an extra layer of privacy, it's not perfect privacy. It doesn't stop someone sending you Bitcoin hoping you mix that Bitcoin with other Bitcoin in your wallet during a spend. That happens already without BIP47. And it costs money to try to violate someone's privacy with that technique. 

P address allows you to make contacts in your wallet for conveniently sending transactions to the same contact but using fresh addresses every time. Making it easier to avoid address reuse means blockchain analysis firms learn less about your transaction history and patterns.

**Comment by SuperPhatArrow** [comment]
FWIW I am very keen to add this feature and I had begun working on it when the previous developers said they were going to do it. There are a few things though...

This issue is for BIP47 version 3 & 4 but this BIP is actually not a BIP as far as I can tell.
[BIP-0047](https://github.com/bitcoin/bips/blob/master/bip-0047.mediawiki) covers versions 1 & 2 
[OBPP-05](https://github.com/OpenBitcoinPrivacyProject/rfc/blob/master/obpp-05.mediawiki) covers versions 3 & 4

In order to be completely compatible with other wallets, like Samourai Wallet and Sparrow Wallet, which both use version 1, Blue wallet should also use version 1 (for now). The only wallet I have heard of that has implemented v3/4 is a wallet called Métier, which is a work-in-progress afaik. It's from Justus Ranvier (the author of OBPP-05).

@yoshimo is there any particular reason for requesting version 3 & 4 when it is not clear that other wallets will be able to send to Blue Wallet if these versions are implemented? Am I correct in assuming that you just wanted the privacy of Payment Codes and BIP-0047 and a version that is compatible with the BIP (and other wallets) rather than the Latest from OBPP?

Then, the next question is for those that pledged: 
@yoshimo 150,000
@librepgp 850,000
@acantril 1,000,000
@dav1dpgit 100,000
[Deleted User] [100,000]
@johnnywonder1 25,000
@coinsurenz 1,000,000
@modl21 1,000,000
@RequestPrivacy 25,000
@Jorj_X_McKie on Twitter via @librepgp  250,000
Citadel Dispatch matrix group user via @librepgp 1,000,000
@Ziya-Sadr 100,000
@apemithrandir  500,000
@btcuserguide 1,000,000
@BTclaudia 500,000
@ion497 200,000
[@Does_not_mattr](https://twitter.com/Does_not_mattr) on Twitter via @librepgp 300,000
@0xAsteria 200,000
@jamesgmorgan 125,000
[Meister Eder](https://twitter.com/Meister_Ancap) on Twitter via @librepgp 100,000
@diegoperezcham 475,000
@TheGuySwann 1,000,000
@alaznem 40,000
@FreddyPPFinklemeyerIII 1,000,000
[Isaiah Douglass, MBA, CFP](https://twitter.com/IDFinancial) on Twitter via @librepgp 500,000
[AmBird](https://twitter.com/TurboSloth1869) on Twitter via @librepgp 300,000
[Honk Honk MetaMick](https://twitter.com/metamick14) on Twitter via @librepgp 100,000
@pointbiz 19,500,000

***TOTAL: 0.3084BTC***

Do your pledges still stand if this is for V1 (compatible) or do you require V3&4 (Probably incompatible) to consider your requirements met?

As for karozagorus, just ignore them. It is their day job to spread FUD about Samourai. Click the links in their profile to see.

**Comment by BitcoinQnA** [comment]
Put me down for 250k sats

**Comment by SuperPhatArrow** [comment]
For reference https://twitter.com/SamouraiDev/status/1499796801733419009

**Comment by btcuserguide** [comment]
Greater compatibility is preferable at this stage so confirming my pledge stands for v1

**Comment by yoshimo** [comment]
Version 1 is just the beginning and has limitations , both technical and in the number of wallets that support it.
Implementing that one would benefit a lot of people but i put the successor on the list in the hope that either we get all of them checked immediately and are one step ahead this time or we at least make the developers aware of the improved draft so that the implementation is prepared and can be enhanced easily later on.

V1 will receive my sats regardless.
Extra sats are on the table if obpp-05 and/or reusable taproot adresses get implemented.

**Comment by BitcoinQnA** [comment]
Additional 50k sats from twitter user [@janetYstackx](https://twitter.com/janetYstackx?s=09)

**Comment by SuperPhatArrow** [comment]
[Isaiah Douglass, MBA, CFP](https://twitter.com/IDFinancial) on Twitter
[Says his pledge still stands](https://twitter.com/IDFinancial/status/1508153202516148224)

**Comment by pointbiz** [comment]
V3 eliminates toxic change. In theory v1 and v3 should be interoperable (at least if Bluewallet has v3 then we can send to a v1). 

Bitcoin has no rulers. BIP process is purely a means of coordination. There's no repository that is authoritative for protocol standards. If one or two developers are impeding BIP47 coordination on an specific URL then we just reference a new URL. This is a necessary response mechanism. 

BIP47 v1 is an improvement to static address reuse which is the exact use case it replaces. BIP47 doesn't in practice create additional blockchain spam because in practice users send a test amount to an address then send the real payment to that same address. This is akin to the connection transaction of BIP47.

I will pay the bounty I pledged for v1/v2 or v3/v4 to be shipped with Bluewallet. 

Samourai lacks USD exchange rates and this makes it hard for beginners (it's hard to use for other reasons too). Bluewallet is a standard bearer and THE Bitcoin wallet therefore the adoption of BIP47 by Bluewallet will cement the standard.

**Comment by SuperPhatArrow** [comment]
>In theory v1 and v3 should be interoperable (at least if Bluewallet has v3 then we can send to a v1).

You are quite right. The problem is the inverse of the second half of the quote, in parentheses. That is that V3 Blue Wallet may not be able to receive from another V1.

So as someone trying to add this to Blue Wallet, I can implement V1 and be certain that Samourai Wallet users and Sparrow Wallet users can pay to the V1 Blue Wallet payment code. If we jump to V3 or V4, Blue wallet users will be able to pay other wallets, but it is not clear if they will be able to receive from them. Although the [spec of OBPP-05](https://github.com/OpenBitcoinPrivacyProject/rfc/blob/master/obpp-05.mediawiki#compatibility-rules) says that versions are backwards compatible, it also says that old versions need to be forwards compatible, but that doesn't mean they will be without changing their code. This part of the spec seems illogical to me, but I will spend more time on it.

So from my point of view there are tough choices to make:
a.) Do V1, like everyone else, and make it interoperable. Once the UI and all is in, changing to V2/3/4 is just a case of updating some functions to update. Side note, updating versions will make a new payment codes (each version produces a different Payment Code as the first byte in the binary payload is the version number) so we can have 4 payment codes, one for each version which seems messy for the user.
b.) Go straight for V4, run the risk of not being able to receive funds.

I think option a.) is best for now with the ability to (hopefully easily) update in future once some of the other wallets are able to support it and some of the complications are better understood.

Please correct me where I am wrong.

**Comment by SuperPhatArrow** [comment]
@modl21 indicated that his pledge stands for V1

**Comment by acantril** [comment]
my pledge stands for V1 upwards, whatever the consensus is (@SuperPhatArrow I appreciate the detail in your post - your logic is sound)

**Comment by TheGuySwann** [comment]
ACK for v1, interoperability matters more than having the “perfect version.”

That said, I would also like to see continuation toward the newer versions as a part of the plan. Maybe I’ll release 750k for V1, and then the other 250k for the next iteration?🤔

Or I might just add an additional donation for the next version. Just getting payment codes operating is really the goal.

**Comment by acantril** [comment]
yeah I'll add more for future enhancements, just getting a interoperable solution for paynyms if my stage 1 focus.

**Comment by chrissyjai** [comment]
Dont put me down for 250 k thanks . Official owner tp this email and all accounts


Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by apemithrandir** [comment]
You have my Sats @SuperPhatArrow 
![YouHaveMyBip47](https://user-images.githubusercontent.com/99969638/160307127-640e3f6a-6eb3-4264-88d4-22784a1e3061.gif)

**Comment by bitcoind-me** [comment]
We are in for 100K sats.

**Comment by RequestPrivacy** [comment]
As I don't use BlueWallet and also don't intend to, my small 25k sats pledge is for the v1 and will be paid to your payment code after BlueWallet users confirm (or I test it myself) that the feature is usable in a sense that people can pay and recife via payment codes.

For future updates of this pledge I will use this

**Comment by RequestPrivacy** [comment]
As I don't use BlueWallet and also don't intend to, my small 25k sats pledge stands for v1 and will be paid to your payment code after BlueWallet users confirm (or I test it myself) that the feature is usable in a sense that people can pay and receive via payment codes.

For future updates of this pledge I will use this message. 
**Pledge stands as of 2022_03_28**

**Comment by linkinparkrulz** [comment]
100,000 sats from me. ACK for V1

**Comment by seniordevel** [comment]
hello
I am react native expert.
I can integrate bluewallet
thanks

**Comment by coinsurenz** [comment]
Sorry I missed this, V1 is acceptable for me and my pledge stands

**Comment by yoshimo** [comment]
@SuperPhatArrow have you made progress on the implementation yet that you can share with us ?

**Comment by SuperPhatArrow** [comment]
@yoshimo sorry for the delayed response.

I have been busy wrapping up previous commitments and deliberately did not send a draft pull request in case someone else could get to this first.

I wrote a BIP47 factory function which could easily be added to any onchain wallet classes generated by the app. I disagree with the way that the previous attempt was going to do it, that is, to make a new kind of wallet called a "Privacy wallet". I think that this gives the impression that other wallet types in the app aren't private. I also think this should be a feature available on *all* on-chain wallets. The code just landed up looking pretty much like the [BIP47 Library](https://github.com/spsina/bip47) that @spsina and @alimahdiyar did.

I had some questions about dependencies that I emailed to bluewallet@bluewallet.io 3 weeks ago but I didn't get a response. The [README.md](https://github.com/BlueWallet/BlueWallet/blob/master/README.md) says to use LTS versions of node (currently v16 & v14) but NPM refuses to install the dependencies on node 16. Using node v14, NPM installs with a lot of warnings about deprecated packages and alerts that the project has `21 vulnerabilities (6 low, 13 high, 2 critical)`. I have not looked into these any further and perhaps they are just dev dependencies. Most of them can be fixed with `npm audit fix`. Nonetheless, this makes me rather uncomfortable.

**To be clear, I am not suggesting that these vulnerabilities impact the security of the app or put users' funds at risk. I don't know that.**

I am sorry to say that I will not be working on Bluewallet. The bounty is generous and would be a lot of money to me but I would rather not be involved. I code in my free time for fun and I have found that trying to work on this app has been extremely frustrating for me. I am not disparaging of the work of any of the previous contributors to the project. I found the code that I looked at to be clear and easy to read. I was, however, still unable to add the feature.

My recommendation is that one of the current contributors that completely grok this codebase use the library that I linked above to add BIP47 functionality to all of Bluewallet's on-chain wallets and share a portion of the bounty with @spsina who lists their paynym in the library readme file. I also hope that the current contributors will take a close look at their dependency tree and update packages that have vulnerabilities.

**Comment by alaznem** [comment]
> V1 (compatible) or do you require V3&4

I'm okay with V1 or V3&4 for my 40k pledge.

**Comment by TheGuySwann** [comment]
Is there any update on this? I would still REALLY like to have paynyms in this app 🙏🏻

**Comment by yoshimo** [comment]
https://github.com/BlueWallet/BlueWallet/commits/limpbrains-paynum-2 has gone stale since july. I still think that there are enough sats on the table to be implemented eventually.

**Comment by TheGuySwann** [comment]
WHY NO PAAAAYNYYYYYYMMMMS? 

Here for my monthly bother 👍🏻

**Comment by abhiShandy** [comment]
I have started working on implementing this. You can follow my [notes](https://abhishandy.notion.site/Implementing-PayNym-in-BlueWallet-1453f97a4c644062b0d76d1e2f1ed874)

**Comment by acantril** [comment]
ooooo.... excited.

**Comment by abhiShandy** [comment]
I am able to generate a Payment code and create a PayNym on BlueWallet. However, BlueWallet doesn't support testnet. Let me know if someone here is willing to spend some sats to test this out on the mainnet. Mempool is also fairly empty right now.

**Comment by linkinparkrulz** [comment]
Yeah, let's do it!

**Comment by abhiShandy** [comment]
Add this Payment code using Samourai or Sparrow wallet:

PM8TJYBZpg1amLvUwe3ZpiLUBCLvsGqLwuahnMciYzdrByGVPDwJ5ZWsaGzEijyES1xK2RyN5j5u6zqR5k6mfzSzR53X6ZB31RFSuonx5W1VU92nEf9P

The paynym is +proudwood3Bb

**Comment by acantril** [comment]
Trying to add it and I get this
<img width="529" alt="Screenshot 2022-12-06 at 11 25 36 am" src="https://user-images.githubusercontent.com/15226029/205785483-b0d1b811-e697-4407-ac03-386d567b54b0.png">

**Comment by acantril** [comment]
I'll look if i can see anything in the logs etc.

**Comment by abhiShandy** [comment]
It should show up since it's visible here https://paynym.is/+proudwood3Bb

**Comment by abhiShandy** [comment]
I have found at least one issue - Bluewallet is not watching BIP47 addresses.

**Comment by pointbiz** [comment]
I'm not sure but I think you have to watch for public keys. I'm also not sure if the Electrum protocol indexes by public keys.

**Comment by ncoelho** [comment]
Development on this is still ongoing you can see the pull request. Just done by open source volunteers, I would suggest increasing the bounty, complaining helps zero.

**Comment by yoshimo** [comment]
I stand by my commitment and every other wallet that implements this also gets sats.

**Comment by ncoelho** [comment]
Bluewallet is done by open source volunteers. It has no full time resources...

**Comment by linkinparkrulz** [comment]
> > Bluewallet is done by open source contributors. It has no full time resources...
> 
> Apologies. I didn't know. The wallet looks so professional I just assumed... edit: @ncoelho Knowing now that it is all done by volunteers, I will up my commitment to 1m sats. I still see the value in PayNyms getting added to BlueWallet and had an uninformed understanding of how BlueWallet was structured.

Consider contributing some of your bounty to the Geyser Fund to get Abhi across the Proof of Concept milestone:

https://geyser.fund/project/bwbip47

**Comment by RequestPrivacy** [comment]
Let us know your payment code :+1:

**Comment by linkinparkrulz** [comment]
> Let us know your payment code +1

@abhiShandy 's payment code is here: https://paynym.is/+smallresonance866

**Comment by abhiShandy** [comment]
that's correct: https://paynym.is/+smallresonance866

**Comment by abhishandilya** [comment]
You can't send using payment codes, just receive from it right now

**Comment by abhiShandy** [comment]
Sorry for the confusion. This is the first iteration. Sending should be implemented in the next one. Having an expiration date is a good idea. This GitHub issue has been lingering for too long.

Thanks for sending half of your pledge.

**Comment by coinsurenz** [comment]
Haven't been able to keep up with developments here but understand this is in beta? once deployed for all and sundry please tag me here & dm/mention on twitter (same handle) if you dont get a prompt response as my github notifications tend to be pretty cluttered

**Comment by jordan-bravo** [comment]
Subscribing to this issue to be notified of updates.

**Comment by abhiShandy** [comment]
That's correct. We can only receive using payment codes.

There has been a few bug fixes and UX improvements on the receiving end.

**Comment by yoshimo** [comment]
Is there ongoing work on the sending part? I'd send my bounty with BW :)

**Comment by mateusdeap** [comment]
Where's the PR for this? Both PRs linked have been closed...

**Comment by pointbiz** [comment]
https://twitter.com/BtcCuracao/status/1783705971363025290

19,500,000 sat bounty from me.

**Comment by Overtorment** [comment]
work is being done to add support for sending to barebone payment codes (not paynyms)

**Comment by pointbiz** [comment]
I see this bounty has been fulfilled and I would like to pay it out. Thank you, from all Canadian Truckers!

I understand the developer was: Overtorment
Reviewers: limpbrains, marcosrdz, BenWestgate

Overtorment can you provide a Bitcoin address or payment code?
Reviewers can you provide Lightning addresses?

**Comment by sara554-beep** [comment]
Hola.
0x4c148F097E8F38B89D3447C8DD4ee37F506bFF39

**Comment by DeMonkeyCoder** [comment]
Hi
Glad to see this feature is finally merged and live!

This [bip47 library](https://github.com/zkfriendly/bip47) developed by @zkfriendly is used in the development of this feature:
https://github.com/BlueWallet/BlueWallet/blob/5cb3aa1b25a58620df1c65df484436ae8c0ead32/class/wallets/abstract-hd-electrum-wallet.ts#L2
So we'd appreciate it if the supporters of the bounty also consider donating some BTC to the donation address provided in [the README](https://github.com/zkfriendly/bip47#donations) of the repository :slightly_smiling_face: 

Thanks for your hard work! @abhiShandy @Overtorment

**Comment by Overtorment** [comment]
i also enquired HRF about the bounty, waiting for a reply. If paid out, it will be distributed to everyone involved, starting from the awesome lib author

**Comment by pointbiz** [comment]
> Hi Glad to see this feature is finally merged and live!
> 
> This [bip47 library](https://github.com/zkfriendly/bip47) developed by @zkfriendly is used in the development of this feature:
> 
> https://github.com/BlueWallet/BlueWallet/blob/5cb3aa1b25a58620df1c65df484436ae8c0ead32/class/wallets/abstract-hd-electrum-wallet.ts#L2
> 
> 
> So we'd appreciate it if the supporters of the bounty also consider donating some BTC to the donation address provided in [the README](https://github.com/zkfriendly/bip47#donations) of the repository 🙂
> Thanks for your hard work! @abhiShandy @Overtorment

Yes! Thank you, I see the address in the readme. My intention is to split my bounty with everyone involved. I will wait to get the addresses from the other developers.

**Comment by yoshimo** [comment]
i would appreciate  payment codes to be listed by the project owners themselfes so nobody can come in and snatch the rewards

**Comment by abhiShandy** [comment]
Here's my payment code:

PM8TJi7Qfijct7b7mZydJkRUY4WPvg9C5UrGL2LZQf9tThsi5RFDTFZorT4DBbzeom9q6SWvSXtw6B9BQkkcSJ7zzNkeKb9KtiQWJZsJ2LscuG5xk9sc

<img height=400 src="https://github.com/user-attachments/assets/b6cdcbef-9103-48f6-9cb7-0aa5f3e1bec5" >

**Comment by Overtorment** [comment]
@zkfriendly can you pls post your paymentcode?

**Comment by zkfriendly** [comment]
Here is my payment code:

PM8TJLSPLroADRS1hJafJEUmRvuXZpE2YmRrfHT1LrYmQW6t4ZxtP73sjh5qUmxMMrpsioFHxaHFpGgkHz5Ubidyzc1orYjcAWYzdA71RVZshdytY67m

-------- Original Message --------

**Comment by DeMonkeyCoder** [comment]
Hi @Overtorment,
Did you also post your payment code somewhere?

**Comment by Overtorment** [comment]
PM8TJTSwtBAgeSZjAG7bAEoRaLkD7cvP8sopgEMpG3fm1bxorfLDRk1w1ih2VxS9RGCstuY4mh1dJJr9ckV4mctvqirq14wgCQcdYHuRG5obe5W5QYus

![image](https://github.com/user-attachments/assets/610bec38-4ecb-4467-bd9b-03968df55a07)

**Comment by Overtorment** [comment]
Some donations were just distributed.

If anyone pledged sats and havent sent them - please split them between 3 listed payment codes (zkfriendly, abhi, myself)

**Comment by pointbiz** [comment]
> Some donations were just distributed.
> 
> If anyone pledged sats and havent sent them - please split them between 3 listed payment codes (zkfriendly, abhi, myself)

I initially pledged 19,500,000 sats. I decided to donate 16,680,000 sats to each developer zkfriendly, abhiShandy, Overtorment totaling 50,040,000 sats.

I'm super thankful to have this contacts functionality and I'm already using it in the real world. You folks are amazing!
